engine/runtime/src/main/java/org/enso/interpreter/runtime/ResourceManager.java

```diff
@@ -234,14 +239,23 @@ public Item(Object underlying, Object finalizer, PhantomReference<ManagedResourc
      * @param context current execution context
      */
     public void doFinalize(EnsoContext context) {
-      Object p = context.getThreadManager().enter();
-      try {
-        InteropLibrary.getUncached(finalizer).execute(finalizer, underlying);
-      } catch (Exception e) {
-        context.getErr().println("Exception in finalizer: " + e.getMessage());
-      } finally {
-        context.getThreadManager().leave(p);
-      }
+      var futureToCancel = new AtomicReference<Future<Void>>(null);
+      var performFinalizer = new ThreadLocalAction(false, false, true) {
+        @Override
+        protected void perform(ThreadLocalAction.Access access) {
+          var tmp = futureToCancel.getAndSet(null);
+          if (tmp == null) {
```

Comment:
Rather than directly executing the `finalizer` we schedule a _"safe point action"_. As soon as it is picked up by some Enso executing thread, we cancel any further executions of the action and execute the _"finalizer"_.